src/main/java/org/elasticsearch/common/geo/builders/BasePolygonBuilder.java

```diff
@@ -112,6 +116,18 @@ public ShapeBuilder close() {
     }
 
     /**
+     * Validates only 1 vertex is tangential (shared) between the interior and exterior of a polygon
+     */
+    protected void validateHole(BaseLineStringBuilder shell, BaseLineStringBuilder hole) {
+        ArrayList exterior = Lists.newArrayList(Sets.newHashSet(shell.points));
```

Comment:
Why the intermediate hash set? The hash set will have random iteration order right? So then the retainAll below becomes O(N^2)? Why not just copy the array lists, sort, and then call retainAll?